stdlib/stdlib-external/runtime/string.ts

```diff
@@ -15,3 +46,292 @@ export function concat(s1: u32, s2: u32): u32 {
+        let oldRet = ret
+        ret = concat(ret, fieldName)
+        decRef(oldRet)
+        oldRet = ret
+        ret = rightLiteralConcat(ret, ': ')
+        decRef(oldRet)
+        oldRet = ret
+        ret = concat(ret, fieldValue)
+        decRef(oldRet)
+        decRef(fieldValue)
+      }
+      decRef(spacePadding)
+      let oldRet = ret
+      ret = rightLiteralConcat(ret, '\n')
+      decRef(oldRet)
+      oldRet = ret
+      ret = concat(ret, lastSpacePadding)
+      decRef(oldRet)
+      oldRet = ret
+      ret = rightLiteralConcat(ret, '}')
+      decRef(oldRet)
+      decRef(lastSpacePadding)
+      return ret
+    }
+    case GRAIN_ARRAY_HEAP_TAG: {
+      let ret = wrapString("[> ")
+      let arity = load<i32>(untaggedPtr, 4 * 1)
+      for (let i = 0; i < arity; ++i) {
+        if (i > 0) {
+          let oldRet = ret
+          ret = rightLiteralConcat(ret, ', ')
+          decRef(oldRet)
+        }
+        let oldRet = ret
+        let tmp = grainToStringHelp(load<u32>(untaggedPtr + 4 * (2 + i)), extraIndents)
+        ret = concat(ret, tmp)
+        decRef(tmp)
+        decRef(oldRet)
+      }
+      let oldRet = ret
+      ret = rightLiteralConcat(ret, ']')
+      decRef(oldRet)
+      return ret
+    }
+    case GRAIN_BOXED_NUM_HEAP_TAG: {
+      let numberTag = load<u32>(untaggedPtr, 4 * 1)
+      switch (numberTag) {
+        case GRAIN_INT32_BOXED_NUM_TAG: {
+          return itoa32(loadI32(untaggedPtr), 10)
+        }
+        case GRAIN_INT64_BOXED_NUM_TAG: {
+          return itoa64(loadI64(untaggedPtr), 10)
+        }
+        case GRAIN_RATIONAL_BOXED_NUM_TAG: {
+          let numerator = loadRationalNumerator(untaggedPtr)
+          let denominator = loadRationalDenominator(untaggedPtr)
+          let ret = itoa32(numerator, 10)
+          let oldRet = ret
+          ret = rightLiteralConcat(ret, '/')
+          decRef(oldRet)
+          oldRet = ret
+          let tmp = itoa32(denominator, 10)
+          ret = concat(ret, tmp)
+          decRef(tmp)
+          decRef(oldRet)
+          return ret
+        }
+        case GRAIN_FLOAT32_BOXED_NUM_TAG: {
+          return dtoa(loadF32(untaggedPtr))
+        }
+        case GRAIN_FLOAT64_BOXED_NUM_TAG: {
+          return dtoa(loadF64(untaggedPtr))
+        }
+      }
+    }
+    default: {
+      let tmp = itoa32(tag, 16)
+      let tmp2 = rightLiteralConcat(tmp, " | value: 0x")
+      let tmp3 = itoa32(ptr, 16)
+      let tmp4 = concat(tmp2, tmp3)
+      let tmp5 = rightLiteralConcat(tmp4, ">")
+      let ret = leftLiteralConcat("<unknown heap tag type: 0x", tmp5)
+      decRef(tmp5)
+      decRef(tmp4)
+      decRef(tmp3)
+      decRef(tmp2)
+      decRef(tmp)
+      return ret
+    }
+  }
+}
+
+function grainToStringHelp(grainValue: u32, extraIndents: u32): u32 {
+  if (!(grainValue & 1)) {
+    // Simple (unboxed) numbers
+    return itoa32(<i32>(grainValue) >> 1, 10)
+  } else if ((grainValue & 7) == GRAIN_TUPLE_TAG_TYPE) {
+    // [TODO]
```

Comment:
What's the TODO here?

Reply:
@ospencer There's _always_ more work TODO.